keps/sig-node/2033-rootless/README.md

```diff
@@ -0,0 +1,806 @@
+
+An external controller should watch changes on `corev1.Service` resources and call [RootlessKit API](https://github.com/rootless-containers/rootlesskit/blob/v0.9.5/pkg/api/openapi.yaml) to set up port forwarding for the node ports.
+
+k3s implementation: https://github.com/rancher/k3s/blob/v1.17.2+k3s1/pkg/rootlessports/controller.go#L92-L96
+
+##### RootlessKit network drivers
+
+RootlessKit supports two kinds of network stacks:
+* TAP with pure usermode network stack (either `slirp4netns` or VPNKit)
+* vEth with setuid binary `lxc-user-nic`
+
+`slirp4netns` is preferred for security, `lxc-user-nic` is preferred for performance.
+
+These stacks are used for the namespace where the node components are executed in, not for the containers' namespaces.
+CNI plugins such as Flannel are expected to be used for the containers' namespace.
+
+##### CNI plugins
+
+Flannel (VXLAN) is known to work.
+
+#### Pod SecurityContext
+
+- `runAsUser`: supported, but the number of the UID is limited by `/etc/subuid`.
+- `sysctls`: some sysctl parameters are supported, but some would fail in `EPERM`.
+  Creating Pod manifests with such sysctl parameters would fail.
+  If this behavior is problematic, user should write a Mutating Admission Webhook to remove such sysctl parameters from Pod manifests.
+- seccomp: supported
+- AppArmor: unsupported. Creating Pod manifests with AppArmor profile specified would fail.
+- [Node-level UserNS KEP](https://github.com/kubernetes/enhancements/issues/127): can be supported. This UserNS will be nested inside the Rootless mode's UserNS.
+
+#### Required changes to Kubernetes
+`kubelet` needs to be updated to ignore errors during [`setupKernelTunables()`](https://github.com/kubernetes/kubernetes/blob/v1.19.0-beta.2/pkg/kubelet/cm/container_manager_linux.go#L389-L428),
+because some of sysctl parameters cannot be modified in the user namespace, and modifying them is not mandatory.
+
+> **Note**
+> These sysctl parameters are set by `kubelet` itself and unrelated to `.spec.securityContext.sysctls` in Pod manifests.
+
+Same applies to the `setRLimit()` and the conntrack stuff in `kube-proxy`.
+
+**Patches**
+- [kubelet/cm: ignore sysctl error when running in userns](https://github.com/rootless-containers/usernetes/blob/v20200701.0/src/patches/kubernetes/0001-kubelet-cm-ignore-sysctl-error-when-running-in-usern.patch)
+- [kube-proxy: allow running in userns](https://github.com/rootless-containers/usernetes/blob/v20200701.0/src/patches/kubernetes/0002-kube-proxy-allow-running-in-userns.patch)
+
+### cgroup
+
+[cgroup v2](../20191118-cgroups-v2.md) and systemd are required. cgroup v1 won't be supported due to security concerns.
+
+containerd supports cgroup v2 for rootless since containerd v1.4.
+
+#### Required changes to Kubernetes
+
+`pkg/kubelet/cm` needs to be updated to set `PodConfig.Linux.CgroupParent` to `"user.slice"` so that CRI/OCI can talk to the user instance of systemd.
```

Comment:
is the qos tree changed?
can you describe how /kubepods/{burstable|besteffort} hierarchy is captured?

Reply:
Please see https://github.com/kubernetes/enhancements/pull/1371#discussion_r523939873

Reply:
Updated KEP to support cgroupsPerQos